File: Examples/GroupDocs.Viewer.Examples.VisualBasic/FtpInputDataHandler.vb

Imports GroupDocs.Viewer.Domain
Imports GroupDocs.Viewer.Domain.Options
Imports GroupDocs.Viewer.Handler.Input
Imports System.Collections.Generic
Imports System.IO
Imports System.Linq
Imports System.Net
Imports System.Text

Namespace GroupDocs.Viewer.Examples
    Public Class FtpInputDataHandler
        Implements IInputDataHandler
        Private Const _server As String = "ftp://localhost" 
        Private Const _userName As String = "anonymous"
        Private Const _userPassword As String = ""


        Public Function GetFileDescription(guid As String) As FileDescription Implements IInputDataHandler.GetFileDescription
            Return New FileDescription(guid)
        End Function

        Public Function GetFile(guid As String) As Stream Implements IInputDataHandler.GetFile
            Dim uri As Uri = GetUriFromGuid(guid)
            Dim request As FtpWebRequest = GetFtpRequest(uri, WebRequestMethods.Ftp.DownloadFile)

            Dim reader As Stream = request.GetResponse().GetResponseStream()
            Dim result As New MemoryStream()

            Dim bytesRead As Integer = 0
            Dim buffer As Byte() = New Byte(2047) {}
            While True
                bytesRead = reader.Read(buffer, 0, buffer.Length)

                If bytesRead = 0 Then
                    Exit While
                End If

                result.Write(buffer, 0, bytesRead)
            End While

            Return result
        End Function

        Public Function GetLastModificationDate(guid As String) As DateTime Implements IInputDataHandler.GetLastModificationDate
            Dim uri As Uri = GetUriFromGuid(guid)
            Dim request As FtpWebRequest = GetFtpRequest(uri, WebRequestMethods.Ftp.GetDateTimestamp)

            Using response As FtpWebResponse = DirectCast(request.GetResponse(), FtpWebResponse)
                Return response.LastModified
            End Using
        End Function
        Public Function LoadFileTree(fileListOptions As FileListOptions) As List(Of FileDescription)
            'TODO
            Return New List(Of FileDescription)()
        End Function
       
        Private Function GetUriFromGuid(guid As String) As Uri
            Return If(Uri.IsWellFormedUriString(guid, UriKind.Absolute), New Uri(guid), New Uri(String.Format("{0}/{1}", _server, guid)))
        End Function

        Private Function GetFtpRequest(uri As Uri, method As String) As FtpWebRequest
            Dim request As FtpWebRequest = DirectCast(WebRequest.Create(uri), FtpWebRequest)
            request.Credentials = New NetworkCredential(_userName, _userPassword)
            request.Method = method
            Return request
        End Function
        Private Shared Function InlineAssignHelper(Of T)(ByRef target As T, value As T) As T
            target = value
            Return value
        End Function
        
        Public Sub AddFile(guid As String, content As Stream) Implements IInputDataHandler.AddFile
            'TODO
        End Sub
        Public Function GetEntities(path As String) As List(Of FileDescription) Implements IInputDataHandler.GetEntities
            'TODO
            Return New List(Of FileDescription)()
        End Function
    End Class
End Namespace

File: Examples/GroupDocs.Viewer.Examples.CSharp/GroupDocs.Viewer.Examples/FtpInputDataHandler.cs
using GroupDocs.Viewer.Domain;
using GroupDocs.Viewer.Domain.Options;
using GroupDocs.Viewer.Handler.Input;
using Microsoft.WindowsAzure.Storage;
using Microsoft.WindowsAzure.Storage.Blob;
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Net;
using System.Text;

namespace GroupDocs.Viewer.Examples.CSharp
{
    public class FtpInputDataHandler : IInputDataHandler
    {
        private const string _server = "ftp://localhost";
        private const string _userName = "anonymous";
        private const string _userPassword = "";

        
        public FileDescription GetFileDescription(string guid)
        {
            return new FileDescription(guid);
        }

        public Stream GetFile(string guid)
        {
            Uri uri = GetUriFromGuid(guid);
            FtpWebRequest request = GetFtpRequest(uri, WebRequestMethods.Ftp.DownloadFile);

            Stream reader = request.GetResponse().GetResponseStream();
            MemoryStream result = new MemoryStream();

            int bytesRead = 0;
            byte[] buffer = new byte[2048];
            while (true)
            {
                bytesRead = reader.Read(buffer, 0, buffer.Length);

                if (bytesRead == 0)
                    break;

                result.Write(buffer, 0, bytesRead);
            }

            return result;
        }

        public DateTime GetLastModificationDate(string guid)
        {
            Uri uri = GetUriFromGuid(guid);
            FtpWebRequest request = GetFtpRequest(uri, WebRequestMethods.Ftp.GetDateTimestamp);

            using (FtpWebResponse response = (FtpWebResponse)request.GetResponse())
                return response.LastModified;
        }

        /// <summary>
        /// Gets files and folders for specified path.
        /// </summary>
        /// <param name="path">The path.</param>
        /// <returns>System.Collections.Generic.List&lt;GroupDocs.Viewer.Domain.FileDescription&gt;.</returns>
        public List<FileDescription> GetEntities(string path)
        {
            Uri uri = Uri.IsWellFormedUriString(path, UriKind.Absolute)
            ? new Uri(path)
            : GetUriFromGuid(path);
            FtpWebRequest request = GetFtpRequest(uri, WebRequestMethods.Ftp.ListDirectory);
            List<FileDescription> result = new List<FileDescription>();
            using (FtpWebResponse response = (FtpWebResponse)request.GetResponse())
            {
                Stream responseStream = response.GetResponseStream();
                if (responseStream != null)
                {
                    using (StreamReader reader = new StreamReader(responseStream))
                    {
                        string guid;
                        while ((guid = reader.ReadLine()) != null)
                        {
                            result.Add(new FileDescription(guid, !guid.Contains(".")));
                        }
                    }
                }
            }
            return result;
        }

        private Uri GetUriFromGuid(string guid)
        {
            return Uri.IsWellFormedUriString(guid, UriKind.Absolute)
                    ? new Uri(guid)
                    : new Uri(string.Format("{0}/{1}", _server, guid));
        }

        private FtpWebRequest GetFtpRequest(Uri uri, string method)
        {
            FtpWebRequest request = (FtpWebRequest)WebRequest.Create(uri);
            request.Credentials = new NetworkCredential(_userName, _userPassword);
            request.Method = method;
            return request;
        }

        public void SaveDocument(CachedDocumentDescription description, Stream stream)
        { 
            //TODO
        }
        /// <summary>
        /// Adds file to storage.
        /// </summary>
        /// <param name="guid">This is user defined key that identifies file in the storage.</param>
        /// <param name="content">Stream to save data to storage.</param>
        public void AddFile(string guid, Stream content)
        {
            //TODO
        }
        public List<FileDescription> LoadFileTree(FileListOptions fileListOptions)
        {
            //TODO
            return new List<FileDescription>();
        }
        [Obsolete]
        public List<FileDescription> LoadFileTree(FileTreeOptions fileTreeOptions)
        {
            //TODO
            return new List<FileDescription>();
        }
    }
}
